src/cargo/ops/vendor.rs

```diff
@@ -387,3 +442,53 @@ fn copy_and_checksum(src_path: &Path, dst_path: &Path, buf: &mut [u8]) -> CargoR
             .with_context(|| format!("failed to write to {:?}", dst_path))?;
     }
 }
+
+/// Maps source id to separate vendor directory name.
+/// E.g. registry+deadbeeffeedcafe, git+0123456789abcdef
+fn source_id_to_dir_name(src_id: SourceId) -> String {
+    let src_type = if src_id.is_registry() {
+        "registry"
+    } else if src_id.is_git() {
+        "git"
+    } else {
+        panic!("Invalid source ID: {src_id}")
+    };
+    struct SourceIdShortHash(SourceId);
+    impl Hash for SourceIdShortHash {
+        fn hash<H: Hasher>(&self, state: &mut H) {
+            self.0.stable_hash(Path::new(""), state);
+        }
+    }
+    let hash = util::short_hash(&SourceIdShortHash(src_id));
```

Comment:
FWIW I think using a different short hash here from `cargo vendor` historically is fine.